### PR TITLE
Replace linkable associations when merging existing array elements for merge patch

### LINKED ISSUE #2596: Merge patch drops linkable associations of existing array elements since 4.3
Since 4.3, a merge-patch (`PATCH` with `application/json`) no longer updates linkable associations of entities that are nested in an array **when the array element already exists**. Non-association properties of the same element are updated, and elements appended beyond the current size (or written into an empty collection) do get their associations resolved. So the very same payload behaves differently depending on the index of the element.

#### Versions

- Observed with Spring Data REST 5.0.5 (Boot 4.0.6), 5.1.1 and current `main` (5.2.0-SNAPSHOT).
- Worked up to 4.2.x. The behavior changed with #2358 (4.3 M1), which switched the handling of existing array elements from `doMerge(…)` to `readPut(…)`.

#### Reproducer

Unit test in the style of `DomainObjectReaderUnitTests` (fails on `main`, the other 42 tests in the class pass):

```java
@Test
void patchReplacesReferenceOfEntityNestedInArray() throws Exception {

	Associations associations = mock(Associations.class);
	when(associations.isLinkableAssociation(any(PersistentProperty.class)))
			.thenAnswer(invocation -> invocation.<PersistentProperty<?>> getArgument(0).isAssociation());
	when(associations.isLinkableAssociation(any(Association.class)))
			.thenAnswer(invocation -> invocation.<Association<?>> getArgument(0).getInverse().isAssociation());

	DomainObjectReader reader = new DomainObjectReader(entities, associations);

	Tag first = new Tag();
	Tag second = new Tag();

	Track track = new Track();
	track.label = "old label";
	track.tag = first;

	Playlist playlist = new Playlist();
	playlist.tracks.add(track);

	SimpleModule module = new SimpleModule().addDeserializer(Tag.class,
			new SelectValueByIdSerializer<Tag>(Map.of(first.id, first, second.id, second)));
	ObjectMapper mapper = JsonMapper.builder().addModule(module).build();

	ObjectNode node = (ObjectNode) mapper.readTree(
			String.format("{ \"tracks\" : [ { \"label\" : \"new label\", \"tag\" : \"%s\" } ] }", second.id));

	Playlist result = reader.doMerge(node, playlist, mapper);

	assertThat(result.tracks).hasSize(1);
	assertThat(result.tracks.get(0).label).isEqualTo("new label"); // passes
	assertThat(result.tracks.get(0).tag).isSameAs(second);         // fails: still `first`
}

@JsonAutoDetect(fieldVisibility = Visibility.ANY)
static class Playlist {
	@Id UUID id = UUID.randomUUID();
	List<Track> tracks = new ArrayList<Track>();
}

@JsonAutoDetect(fieldVisibility = Visibility.ANY)
static class Track {
	String label;
	@Reference Tag tag;
}
```

(`Tag` and `SelectValueByIdSerializer` are the existing fixtures from the DATAREST-1030 test. `Playlist` and `Track` need to be registered with the `KeyValueMappingContext` in `setUp()`.)

The same happens end-to-end with JPA: an `@ElementCollection` of an `@Embeddable` that holds a `@ManyToOne` to an exported entity. `PATCH /orders/{id}` with `{"items":[{"quantity":2,"product":"/products/{id}"}]}` changes `quantity` but leaves `product` untouched, while `PATCH {"items":[]}` followed by the same request applies both.

#### Analysis

`DomainObjectReader#handleArrayNode` applies an `ObjectNode` to an existing element via `readPut(…)`:

```java
if (ObjectNode.class.isInstance(jsonNode)) {
	nestedObjectFound = true;
	readPut((ObjectNode) jsonNode, next, mapper);
}
```

`readPut` deserializes the node into a fresh instance and calls `mergeForPut(source, target, mapper)`, which copies properties with `MergingPropertyHandler` but wraps it in `LinkedAssociationSkippingAssociationHandler`, i.e. linkable associations are deliberately not copied. That rule makes sense for a top-level `PUT`, where associations are managed through association resources. For an element nested in an array it does not:

- there is no association resource for a nested element, so the request body is the only place the association can be expressed;
- RFC 7386, which #2358 refers to, treats the array as a value that is replaced as a whole, so the association is part of the replaced value;
- appended elements (`rawValues.apply(current)`) and elements written into an empty collection are deserialized by Jackson and do resolve the association, and top-level association collections are resolved as well (DATAREST-1030). Only the "existing element at the same index" path drops it.

Afterwards the array is removed from the root node (`i.remove()`), so the outer `readerForUpdating(target)` never sees it and cannot recover the association.

#### Proposal

Treat an existing element the same way as an appended one: deserialize the node into a new element (as `rawValues` already does) and replace the existing element in the collection, instead of merging with PUT semantics. This keeps the RFC 7386 replacement semantics introduced in #2358 (`handlesEntityNestedInAnArrayLikePutForPatchRequest` still passes, unspecified properties become `null`) while restoring the association handling.

I'm happy to submit a PR with the test above and the change to `handleArrayNode`.

### PR DESCRIPTION
**Pull request by ogawa-takeshi**
Closes #2596.

Since #2358, an object nested in an array is applied to the existing element at the same index via `readPut(…)` to obtain PUT semantics (RFC 7386). `mergeForPut(…)` skips linkable associations, which is intended for top-level `PUT` requests where associations are managed through association resources. Nested elements have no association resource, so the request body is the only place to express them. As a result, non-association properties of an existing element were updated while its associations kept their previous values, whereas appended elements and elements written into an empty collection resolved them properly.

This change

- merges existing array elements through a dedicated path (`mergeArrayElement`) that forwards linkable associations to `MergingPropertyHandler`,
- makes `MergingPropertyHandler` replace a linkable association with the reference contained in the payload instead of merging the source state into the referenced instance,
- leaves the top-level `PUT` path (`readPut` / `mergeForPut`) unchanged,
- adds `patchReplacesReferenceOfEntityNestedInArray` to `DomainObjectReaderUnitTests` (fails on `main`, passes with this change; `handlesEntityNestedInAnArrayLikePutForPatchRequest` from #2358 keeps passing).

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/main/CONTRIBUTING.adoc).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/main/etc/ide) and have them applied to your changes. Don't submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/json/DomainObjectReader.java [hunk 1]
```diff
@@ -161,6 +162,38 @@ private void retainIdentifierAndVersion(ObjectNode source, Object target, Object
...
+	/**
+	 * Reads the given {@link ObjectNode} into a new instance of the given existing array element's type and merges it
+	 * into the existing element applying {@literal PUT} semantics. In contrast to {@link #readPut(ObjectNode, Object,
+	 * ObjectMapper)}, linkable associations are replaced by the ones contained in the payload as nested elements are not
+	 * exposed as association resources and the payload is the only place to express them.
+	 *
+	 * @param source must not be {@literal null}.
+	 * @param target must not be {@literal null}.
+	 * @param mapper must not be {@literal null}.
+	 * @throws Exception
```
Comment (ruthst00) [inline spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/json/DomainObjectReader.java:177]: This `@throws Exception` could be more descriptive (e.g., `@throws Exception if deserialization of the source node fails`).

File: spring-data-rest-webmvc/src/test/java/org/springframework/data/rest/webmvc/json/DomainObjectReaderUnitTests.java [hunk 2]
```diff
@@ -1118,6 +1154,20 @@ static class Tag {
 		String name;
 	}
 
+	// GH-2596
+
```
Comment (ruthst00) [inline spring-data-rest-webmvc/src/test/java/org/springframework/data/rest/webmvc/json/DomainObjectReaderUnitTests.java:1158]: The comment `// GH-2596` above the new test classes in the test file is a bit loose — it's not attached to a specific annotation. The `@Test` method already has `// GH-2596` in its annotation comment, so the standalone comment above the model classes is slightly redundant.

File: spring-data-rest-webmvc/src/test/java/org/springframework/data/rest/webmvc/json/DomainObjectReaderUnitTests.java [hunk 3]
```diff
@@ -1118,6 +1154,20 @@ static class Tag {
...
+	@JsonAutoDetect(fieldVisibility = Visibility.ANY)
+	static class Playlist {
+		@Id UUID id = UUID.randomUUID();
+		List<Track> tracks = new ArrayList<Track>();
+	}
+
+	@JsonAutoDetect(fieldVisibility = Visibility.ANY)
+	static class Track {
```
Comment (ruthst00) [inline spring-data-rest-webmvc/src/test/java/org/springframework/data/rest/webmvc/json/DomainObjectReaderUnitTests.java:1166]: Add `@Id` on `Track` test class for realism.

File: spring-data-rest-webmvc/src/test/java/org/springframework/data/rest/webmvc/json/DomainObjectReaderUnitTests.java [hunk 4]
```diff
@@ -73,6 +73,7 @@
  * @author Mathias Düsterhöft
  * @author Ken Dombeck
  * @author Thomas Mrozinski
+ * @author Takeshi Ogawa
  */
 @ExtendWith(MockitoExtension.class)
 class DomainObjectReaderUnitTests {
```
Comment (ruthst00) [inline spring-data-rest-webmvc/src/test/java/org/springframework/data/rest/webmvc/json/DomainObjectReaderUnitTests.java:79]: Add edge-case tests for `null` association and appended elements:

- For example, There's no test verifying that sending `null` for a linkable association in a nested array element correctly nulls out the reference. This edge case is worth covering.
- Also, there's no test for the case where the array grows (new element with a linkable association).
- The PR description mentions "appended elements and elements written into an empty collection resolved them properly" — implying those paths already worked. It would be good to have a regression test confirming that appending a new element with a linkable association still works correctly after this change (since the append path goes through `rawValues.apply(current)` / `mapper.treeToValue`, not `mergeArrayElement`).

File: spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/json/DomainObjectReader.java [hunk 5]
```diff
@@ -161,6 +162,38 @@ private void retainIdentifierAndVersion(ObjectNode source, Object target, Object
...
 	@Nullable
 	<T> T mergeForPut(T source, T target, final ObjectMapper mapper) {
+		return mergeForPut(source, target, mapper, false);
+	}
+
+	/**
+	 * Reads the given {@link ObjectNode} into a new instance of the given existing array element's type and merges it
+	 * into the existing element applying {@literal PUT} semantics. In contrast to {@link #readPut(ObjectNode, Object,
+	 * ObjectMapper)}, linkable associations are replaced by the ones contained in the payload as nested elements are not
+	 * exposed as association resources and the payload is the only place to express them.
```
Comment (ruthst00) [inline spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/json/DomainObjectReader.java:172]: Javadoc for `mergeArrayElement` could be more precise about what it omits vs. `readPut`.

**Review by ruthst00 — CHANGES_REQUESTED** [review]
